plugins/templates/create.js

```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const boom = require('boom');
+const schema = require('screwdriver-data-schema');
+const hoek = require('hoek');
+const urlLib = require('url');
+
+module.exports = () => ({
+    method: 'POST',
+    path: '/templates',
+    config: {
+        description: 'Create a new template',
+        notes: 'Create a specific template',
+        tags: ['api', 'templates'],
+        auth: {
+            strategies: ['token', 'session'],
+            scope: ['build']
+        },
+        plugins: {
+            'hapi-swagger': {
+                security: [{ token: [] }]
+            }
+        },
+        handler: (request, reply) => {
+            const pipelineFactory = request.server.app.pipelineFactory;
+            const templateFactory = request.server.app.templateFactory;
+            const token = request.auth.credentials;
+            const pipelineId = token.pipelineId;
+            const name = request.payload.name;
+            const version = request.payload.version;
+            const labels = request.payload.labels || [];
+
+            return Promise.all([
+                pipelineFactory.get(pipelineId),
+                templateFactory.list({ name })
+            ]).then(([pipeline, templates]) => {
+                const templateConfig = hoek.applyToDefaults(request.payload, {
+                    scmUri: pipeline.scmUri,
+                    labels
+                });
+
+                // If template doesn't exist yet, just create a new entry
+                if (templates.length === 0) {
+                    return templateFactory.create(templateConfig);
+                }
+
+                // If template exists, but this build's scmUri is not the same as template's scmUri
+                // Then this build does not have permission to publish
+                if (pipeline.scmUri !== templates[0].scmUri) {
+                    throw boom.unauthorized('Not allowed to publish this template');
+                }
+
+                // If template exists and has good permission, check the exact version
+                return templateFactory.get({ name, version })
+                    .then((template) => {
+                        // If the version doesn't exist, create a new entry
```

Comment:
From a REST perspective, I would have expected a `PUT` for being able to specifically create a resource. Looking at how the endpoint responds to the same input, I would also expect it to be a `PUT` because it's an idempotent operation.

For instance, with jobs or builds, we allow you to update the resource directly with a `PUT`. By specifying which build or job (e.g. `/builds/{id}` or `/jobs/{id}`), I can update them.